ESQL: Fix projection generation when pruning left join

bpintea opened the pull request:
Currently, in case the join key is (locally) null, the whole join is replaced with an `Eval` and a `Project`. The project however contain only reference attributes. This is insufficient, in case the attribute should point to another attribute pointing to a literal `null` added in the `Eval`.

This fixes that by allowing the generated `Project` to contain `NamedExpression`s, which can then be proper `Alias`es.
This is done by adjusting the join output calculation to work with the `NamedExpression`s, instead of just `Attribute`s.

bpintea committed:
Fix projection generation when prunning left join

Currently, in case (locally) the join key is null, the whole join is
replaced with an Eval and a Project. The project however contain only
reference attributes. This is insufficient, in case the attribute should
point to another attribute pointing to a literal null added in the Eval.

This fixes that by allowing the generated Project to contain
NamedExpressions, which can then be proper Aliases.

bpintea committed:
Add one more (rest) test

bpintea committed:
Update docs/changelog/135446.yaml

astefan reviewed (approved):
LGTM
For someone like me (who takes a PR for a spin with a sample query and dissects its internals and results) I would like to also see a concrete query in csv-spec file. You shared one with me offline and I think it wouldn't hurt to also have that one bundled in the PR, as well.
`from languages_lookup_non_unique_key | eval language_code = null::integer | lookup join languages_lookup_non_unique_key on language_code`

astefan commented:
Actually... (sorry about that), I had an idea for a test and it turns out we don't have a test for that and with this PR it has a wrong behavior:
`from languages_lookup_non_unique_key | keep country, language_name | eval language_code = null::integer | inline stats max(language_code) by language_code`

Before this PR, the result is:

```
        country         | language_name |max(language_code)| language_code 
------------------------+---------------+------------------+---------------
Canada                  |English        |null              |null           
null                    |English        |null              |null           
United Kingdom          |null           |null              |null           
...
```

with this PR, the order of the columns changes (and it's wrong):

```
        country         | language_name | language_code |max(language_code)
------------------------+---------------+---------------+------------------
Canada                  |English        |null           |null              
null                    |English        |null           |null              
United Kingdom          |null           |null           |null              
...
```

bpintea committed:
Merge remote-tracking branch 'upstream/main' into fix/projection_flattening

bpintea committed:
Fix InlineJoin output merging

bpintea committed:
Merge branch 'main' into fix/projection_flattening

bpintea commented:
> I had an idea for a test and it turns out we don't have a test for that and with this PR it has a wrong behavior

Indeed, due to the updated method overriding, the `InlineJoin` ended up using the wrong method. Thanks for catching it, @astefan! I've updated that and added in the test.

astefan reviewed (approved):
LGTM

bpintea commented:
Thanks, Andrei!
